From 96f87999c795f12be3d1561455f39eab3d361086 Mon Sep 17 00:00:00 2001
From: Gareth Davidson <gaz@bitplane.net>
Date: Sun, 6 Sep 2026 08:48:51 +0100
Subject: [PATCH 2/6] Parse and encode x86 operand prefixes

Carry operand and address widths through instruction selection and prefix
emission. Mark fixed-width instruction forms explicitly so their encoding
uses the requested width in each assembly mode.
---
 Src/AsmX64.HC  | 29 ++++++++++++++++++++++-------
 Src/UAsmX64.HC | 30 +++++++++++++++---------------
 2 files changed, 37 insertions(+), 22 deletions(-)

diff --git a/Src/AsmX64.HC b/Src/AsmX64.HC
index 6116b1c..cefaffb 100644
--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -111,6 +111,20 @@ CX86AddrMode *PrsX86AddrMode(CCmpCtrl *cc) {
   mode->imm=0;
   mode->label=NULL;
 prefix:
+  if(cc->token==TK_IDENT&&cc->hash_entry&&
+        cc->hash_entry->type&HTT_CLASS&&
+        !StrCmp(cc->cur_str,cc->hash_entry->str)) {
+    lookahead=cc->cur_buf_ptr;
+    while(*lookahead&&*lookahead<=' ') lookahead++;
+    tmpc=cc->hash_entry;
+    if(!(tmpc->member_lst_and_root&&
+          (cc->flags&CCF_USE_LAST_U16&&cc->last_U16=='.'||*lookahead=='.'))) {
+      mode->operand_sz=tmpc->size;
+      Lex(cc);
+      if(cc->token!='[')
+        goto size_prefix_consumed;
+    }
+  }
   if(cc->token==TK_IDENT) {
     tmpc=HashFind(cc->cur_str,cmp.asm_hash,HTT_INTERNAL_TYPE);
     if(tmpc) {
@@ -119,8 +133,7 @@ prefix:
       goto size_prefix_consumed;
     }
   }
-  if(cc->token==TK_IDENT||cc->hash_entry&&
-        cc->hash_entry->type&HTT_INTERNAL_TYPE) {
+  if(cc->token==TK_IDENT) {
     lookahead=cc->cur_buf_ptr;
     while(*lookahead&&*lookahead<=' ') lookahead++;
     rg=HashFind(cc->cur_str,cmp.asm_hash,HTT_REG);
@@ -158,6 +171,7 @@ prefix:
       mode->operand_sz=tmpc->size;
       Lex(cc);
     }
+  }
 size_prefix_consumed:
     if(!rg&&cc->token==TK_IDENT)
       rg=HashFind(cc->cur_str,cmp.asm_hash,HTT_REG);
@@ -178,7 +192,6 @@ size_prefix_consumed:
       mode->type=X86MT_REG;
       goto ret;
     }
-  }
 indir:
   if(cc->token=='[') {
     Lex(cc);
@@ -1020,6 +1033,10 @@ rex|=0x40;
     to[len++]=0x66;
   }
 //Prefix group 4(Address size ovveride)
+  if (seg_size==16&&indir_mode ||
+        seg_size==64&&indir_mode&&indir_mode->short_addr==1)
+    to[len++]=0x67;
+
   if(rex&0x40) {
     to[len++]=rex;    
   }
@@ -1352,10 +1369,8 @@ enter:
   cc->flags|=CCF_CLASS_DOL_OFFSET;
   cc->class_dol_offset=len;
   if(cc->token==TK_IDENT) {
-    asm_entry=cc->hash_entry;
-    if(!asm_entry||!(asm_entry->type&(HTT_OPCODE|HTT_ASM_KEYWORD)))
-      asm_entry=HashFind(cc->cur_str,cmp.asm_hash,
-            HTT_OPCODE|HTT_ASM_KEYWORD);
+    asm_entry=HashFind(cc->cur_str,cmp.asm_hash,
+          HTT_OPCODE|HTT_ASM_KEYWORD);
     if(asm_entry) cc->hash_entry=asm_entry;
   }
   MemSetU64(modes,NULL,4);
diff --git a/Src/UAsmX64.HC b/Src/UAsmX64.HC
index 8a61628..2e802df 100644
--- a/Src/UAsmX64.HC
+++ b/Src/UAsmX64.HC
@@ -1110,7 +1110,7 @@ U0 UAsmInit() {
   AO("CALL",-1,-1,0xff,-1,X86_OPCF_MODRM,X86_OPERAND_RMx,X86_OPERAND_EXT,2);
   AO("CALLF",-1,-1,0xff,-1,X86_OPCF_MODRM,X86_OPERAND_RMx,X86_OPERAND_EXT,3);
 
-  AO("CBW",0x66,-1,0x98,-1,0);
+  AO("CBW",-1,-1,0x98,-1,X86_OPCF_OPSIZE16);
   AO("CWDE",-1,-1,0x98,-1,0);
   AO("CDQE",-1,-1,0x98,-1,X86_OPCF_REXW);
 
@@ -1151,7 +1151,7 @@ U0 UAsmInit() {
   AO("CMPPS",-1,0xf,0xc2,-1,X86_OPCF_MODRM,X86_OPERAND_RXMM,X86_OPERAND_RMXMM128,X86_OPERAND_IMM8);
 
   AO("CMPSB",-1,-1,0xa6,-1,0);
-  AO("CMPSW",0x66,-1,0xa7,-1,0);
+  AO("CMPSW",-1,-1,0xa7,-1,X86_OPCF_OPSIZE16);
   AO("CMPSD",-1,-1,0xa7,-1,0);
   AO("CMPSQ",-1,-1,0xa7,-1,X86_OPCF_REXW);
 
@@ -1191,7 +1191,7 @@ U0 UAsmInit() {
   AO("CVTTPS2PI",-1,0xf,0x2c,-1,X86_OPCF_MODRM,X86_OPERAND_RXMM,X86_OPERAND_RMXMM64);
   AO("CVTTSD2SI",0xf2,0xf,0x2c,-1,X86_OPCF_MODRM,X86_OPERAND_Rx,X86_OPERAND_RMXMM64);
   AO("CVTTSS2SI",0xf3,0xf,0x2c,-1,X86_OPCF_MODRM,X86_OPERAND_Rx,X86_OPERAND_RMXMM32);
-  AO("CWD",0x66,-1,0x99,-1,0);
+  AO("CWD",-1,-1,0x99,-1,X86_OPCF_OPSIZE16);
   AO("CDQ",-1,-1,0x99,-1,0);
   AO("CQO",-1,-1,0x99,-1,X86_OPCF_REXW);
   AO("DEC",-1,-1,0xfe,-1,X86_OPCF_MODRM,X86_OPERAND_RM8,X86_OPERAND_EXT,1);
@@ -1360,15 +1360,15 @@ U0 UAsmInit() {
   AO("IMUL2",-1,0xf,0xaf,-1,X86_OPCF_MODRM,X86_OPERAND_Rx,X86_OPERAND_RMx);
 
   AO("IN",-1,-1,0xe4,-1,0,X86_OPERAND_CONST_R8,REG_RAX,X86_OPERAND_IMM8);
-  AO("IN",-1,-1,0xe5,-1,0,X86_OPERAND_CONST_R32,REG_RAX,X86_OPERAND_IMM8);
+  AO("IN",-1,-1,0xe5,-1,X86_OPCF_OPSIZE32,X86_OPERAND_CONST_R32,REG_RAX,X86_OPERAND_IMM8);
   AO("IN",-1,-1,0xec,-1,0,X86_OPERAND_CONST_R8,REG_RAX,X86_OPERAND_CONST_R16,REG_RDX);
-  AO("IN",-1,-1,0xed,-1,0,X86_OPERAND_CONST_R32,REG_RAX,X86_OPERAND_CONST_R16,REG_RDX);
+  AO("IN",-1,-1,0xed,-1,X86_OPCF_OPSIZE32,X86_OPERAND_CONST_R32,REG_RAX,X86_OPERAND_CONST_R16,REG_RDX);
 
   AO("INC",-1,-1,0xfe,-1,X86_OPCF_MODRM,X86_OPERAND_RM8,X86_OPERAND_EXT,0);
   AO("INC",-1,-1,0xff,-1,X86_OPCF_MODRM,X86_OPERAND_RMx,X86_OPERAND_EXT,0);
 
   AO("INSB",-1,-1,0x6c,-1,0);
-  AO("INSW",0x66,-1,0x6d,-1,0);
+  AO("INSW",-1,-1,0x6d,-1,X86_OPCF_OPSIZE16);
   AO("INSD",-1,-1,0x6d,-1,0);
 
   AO("INSERTPS",0x66,0xf,0x3a,0x21,X86_OPCF_MODRM,X86_OPERAND_RXMM,X86_OPERAND_RMXMM32,X86_OPERAND_IMM8);
@@ -1435,7 +1435,7 @@ U0 UAsmInit() {
   AO("LLDT",-1,-1,0x0f,0x00,X86_OPCF_MODRM,X86_OPERAND_Mx,X86_OPERAND_EXT,2);
   AO("LMSW",-1,-1,0x0f,0x01,X86_OPCF_MODRM,X86_OPERAND_Mx,X86_OPERAND_EXT,6);
   AO("LODSB",-1,-1,0xac,-1,0);
-  AO("LODSW",0x66,-1,0xad,-1,0);
+  AO("LODSW",-1,-1,0xad,-1,X86_OPCF_OPSIZE16);
   AO("LODSD",-1,-1,0xad,-1,0);
   AO("LODSQ",-1,-1,0xad,-1,X86_OPCF_REXW);
 
@@ -1470,8 +1470,8 @@ U0 UAsmInit() {
 
   AO("MOV",-1,-1,0xb0,-1,X86_OPCF_ADD_REG,X86_OPERAND_R8,X86_OPERAND_IMM8);
   AO("MOV",-1,-1,0xb8,-1,X86_OPCF_ADD_REG|X86_OPCF_REXW,X86_OPERAND_R64,X86_OPERAND_IMM64);
-  AO("MOV",0x66,-1,0xb8,-1,X86_OPCF_ADD_REG,X86_OPERAND_R16,X86_OPERAND_IMM16);
-  AO("MOV",-1,-1,0xb8,-1,X86_OPCF_ADD_REG,X86_OPERAND_R32,X86_OPERAND_IMM32);
+  AO("MOV",-1,-1,0xb8,-1,X86_OPCF_ADD_REG|X86_OPCF_OPSIZE16,X86_OPERAND_R16,X86_OPERAND_IMM16);
+  AO("MOV",-1,-1,0xb8,-1,X86_OPCF_ADD_REG|X86_OPCF_OPSIZE32,X86_OPERAND_R32,X86_OPERAND_IMM32);
   AO("MOV",-1,-1,0xc6,-1,X86_OPCF_MODRM,X86_OPERAND_RM8,X86_OPERAND_IMM8,X86_OPERAND_EXT,0);
   AO("MOV",-1,-1,0xc7,-1,X86_OPCF_MODRM,X86_OPERAND_RMx,X86_OPERAND_IMMx,X86_OPERAND_EXT,0);
 /*
@@ -1494,7 +1494,7 @@ U0 UAsmInit() {
   AO("MOVDDUP",0xf2,0xf,0x12,-1,X86_OPCF_MODRM,X86_OPERAND_RXMM,X86_OPERAND_RMXMM64);
   AO("MOVDQ2Q",0xf2,0xf,0xd6,-1,X86_OPCF_MODRM,X86_OPERAND_RMM,X86_OPERAND_RMXMM64);
   AO("MOVSB",-1,-1,0xa4,-1,0);
-  AO("MOVSW",0x66,-1,0xa5,-1,0);
+  AO("MOVSW",-1,-1,0xa5,-1,X86_OPCF_OPSIZE16);
   AO("MOVSD",-1,-1,0xa5,-1,0);
   AO("MOVSQ",-1,-1,0xa5,-1,X86_OPCF_REXW);
   AO("REP_MOVSB",-1,-1,0xa4,-1,X86_OPCF_REP);
@@ -1581,11 +1581,11 @@ U0 UAsmInit() {
   AO("ORPD",0x66,0xf,0x56,-1,X86_OPCF_MODRM,X86_OPERAND_RXMM,X86_OPERAND_RMXMM128);
   AO("ORPS",-1,0xf,0x56,-1,X86_OPCF_MODRM,X86_OPERAND_RXMM,X86_OPERAND_RMXMM128);
   AO("OUT",-1,-1,0xe6,-1,0,X86_OPERAND_IMM8,X86_OPERAND_CONST_R8,REG_RAX);
-  AO("OUT",-1,-1,0xe7,-1,0,X86_OPERAND_IMM8,X86_OPERAND_CONST_R32,REG_RAX);
+  AO("OUT",-1,-1,0xe7,-1,X86_OPCF_OPSIZE32,X86_OPERAND_IMM8,X86_OPERAND_CONST_R32,REG_RAX);
   AO("OUT",-1,-1,0xee,-1,0,X86_OPERAND_CONST_R16,REG_RDX,X86_OPERAND_CONST_R8,REG_RAX);
-  AO("OUT",-1,-1,0xef,-1,0,X86_OPERAND_CONST_R16,REG_RDX,X86_OPERAND_CONST_R32,REG_RAX);
+  AO("OUT",-1,-1,0xef,-1,X86_OPCF_OPSIZE32,X86_OPERAND_CONST_R16,REG_RDX,X86_OPERAND_CONST_R32,REG_RAX);
   AO("OUTSB",-1,-1,0x6e,-1,0);
-  AO("OUTSW",0x66,-1,0x6f,-1,0);
+  AO("OUTSW",-1,-1,0x6f,-1,X86_OPCF_OPSIZE16);
   AO("OUTSD",-1,-1,0x6f,-1,0);
 
   AO("PACKSSDW",-1,0xf,0x6b,-1,X86_OPCF_MODRM,X86_OPERAND_RMM,X86_OPERAND_RMMM64);
@@ -1879,7 +1879,7 @@ U0 UAsmInit() {
   AO("SETZ,SETE",-1,0xf,0x94,-1,X86_OPCF_MODRM,X86_OPERAND_RM8,X86_OPERAND_EXT,0);
 
   AO("SCASB",-1,-1,0xae,-1,0);
-  AO("SCASW",0x66,-1,0xaf,-1,0);
+  AO("SCASW",-1,-1,0xaf,-1,X86_OPCF_OPSIZE16);
   AO("SCASD",-1,-1,0xaf,-1,0);
   AO("SCASQ",-1,-1,0xaf,-1,X86_OPCF_REXW);
 
@@ -1923,7 +1923,7 @@ U0 UAsmInit() {
   AO("STI",-1,-1,0xfb,-1,0);  
 //STMXCSR
   AO("STOSB",-1,-1,0xaa,-1,0);
-  AO("STOSW",0x66,-1,0xaa,-1,0);
+  AO("STOSW",-1,-1,0xaa,-1,X86_OPCF_OPSIZE16);
   AO("STOSD",-1,-1,0xab,-1,0);
   AO("STOSQ",-1,-1,0xab,-1,X86_OPCF_REXW);
 //STR
-- 
2.53.0
